doc/superbuild/source/developers/style_guide/SourceCode.rst

```diff
@@ -58,12 +52,312 @@ namespace prefix) should be used for all public function names that are not
+new derived class, use the same style as above for implementations of the base
+class method i.e., ``<short class name><method>_<implementation>``.
+
+.. _Style.Naming.NewClasses:
+
+Names for New Classes
+---------------------
+
+All new base classes should use the naming convention ``<class name>_<method>``
+for the base class methods. See
+:numref:`Style.Table.NewBaseClassMethodNaming` for examples.
+
+.. _Style.Table.NewBaseClassMethodNaming:
+
+.. Table:: SUNDIALS naming conventions for methods in new base classes.
+
+   +-----------------------+------------+---------------------------+
+   | Base Class            | Operation  | Method                    |
+   +-----------------------+------------+---------------------------+
+   | ``SUNMemoryHelper``   | Alloc      | ``SUNMemoryHelper_Alloc`` |
+   +-----------------------+------------+---------------------------+
+
+Derived class implementations of the base class methods should follow the naming
+convention  ``<class name>_<method>_<implementation>``. See
+:numref:`Style.Table.NewDerivedClassMethodNaming` for examples.
+
+.. _Style.Table.NewDerivedClassMethodNaming:
+
+.. Table:: SUNDIALS naming conventions for derived class implementations of
+           methods in new base classes.
+
+   +---------------+---------------------------+--------------------------------+
+   | Derived Class | Base Class Method         | Method Implementation          |
+   +---------------+---------------------------+--------------------------------+
+   | CUDA          | ``SUNMemoryHelper_Alloc`` | ``SUNMemoryHelper_Alloc_Cuda`` |
+   +---------------+---------------------------+--------------------------------+
+
+For destructor functions, use ``Destroy`` rather than ``Free`` or some other alternative.
+
+
+.. _Style.Classes.Cpp:
+
+Naming Convention for C++ Classes
+---------------------------------
+
+C++ classes should have a descriptive name. The class name should not be
+prefixed with ``SUN``, but it should reside in the ``sundials::`` namespace.
+Public C++ class functions should use Pascal case (e.g. ``DoSomething``).
+Private C++ class functions should use camelcase (e.g. ``doSomething``).
+
+C++ private class members should use snake case with a trailing underscore
+(e.g. ``some_var_``).
+
+.. _Style.Code:
+
+Coding Conventions and Rules 
+============================
+
+#. Do not use language features that are not compatible with C99, C++14,
+   and MSVC v1900+ (Visual Studio 2015). Examples of such features include
+   variable-length arrays. Exceptions are allowed when interfacing with a
+   library which requires a newer standard.
+
+#. All new code added to SUNDIALS should be linted with  `clang-tidy
+   <https://clang.llvm.org/extra/clang-tidy/>`_ and formatted with `clang-format
+   <https://clang.llvm.org/docs/ClangFormat.html>`_. See :ref:`Style.Formatting`
+   for details.
+
+#. Split very long lines when it improves code readability by using ``//``.
```

Comment:
Is `//` really a valid approach for splitting very long lines in all SUNDIALS-relevant languages?  Can't this just say:
```suggestion
#. Split very long lines when it improves code readability.
```

Reply:
This rule needs further clarification, and probably an example. Sometimes (mostly in C++ code) you may have code like this:
```c++
MyClass::callAFunctionOfSorts::doSomething().doAnotherThing().doSomethingElse();
```
That you would like to format as (for readability):
```c++
MyObject::callAFunctionOfSorts()
        .doSomething()
        .doAnotherThing()
        .doSomethingElse();
```
Clang-format will not accept the desired form without doing this
```c++
MyObject::callAFunctionOfSorts()
        .doSomething()       \\
        .doAnotherThing()    \\
        .doSomethingElse();  \\
```

Reply:
I think this PR actually already would impose C99 through the error checks in the inline functions in header files.

Reply:
Also, AFAIK there is no standard guarantee of ABI compatibility between C89 and C99. It might be that in most cases they are ABI compatible, but its not required by the standard. So even if our headers are C89/C90 there is no guarantee we won't impose C99 on a code that normally uses C89/C90.

Reply:
You're right the inline function with error handling will impose C99 on the user code as well. Requiring C99 does not seem unreasonable but we should discuss tomorrow.